### PR TITLE
Safe envelope conversion

### PR DESCRIPTION
**Pull request by HendrikPeilke**
As discussed on the geotools-devel mailing list (topic "Envelope conversion loses emptyness"), this pull request tries to extend ReferencedEvenlope implementation, such that emptiness property survives the conversion between different envelope formats.

This pull request includes:
- the changed source code
- some minor fixes in source code (spelling mistakes, elimination of doubly defined methods, unused variables)
- a test

GeoTools and GeoServer build fine on my machine

### CODE REVIEW

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java [hunk 1]
```diff
@@ -776,14 +786,14 @@ public String toString() {
     /**
```
Comment (HendrikPeilke) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: just corrected a spelling mistake down here

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java [hunk 2]
```diff
@@ -875,21 +897,7 @@ public static ReferencedEnvelope reference(Envelope e) {
             return new ReferencedEnvelope(e, null);
         }
     }
-
```
Comment (HendrikPeilke) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: method is a duplicate to reference(org.opengis.geometry.Envelope env), which is not needed

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java [hunk 3]
```diff
@@ -925,10 +935,36 @@ public static ReferencedEnvelope reference(org.opengis.geometry.Envelope env) {
             return (ReferencedEnvelope) env;
         }
     
-        if (env.getDimension() == 3) {
```
Comment (HendrikPeilke) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: changed to >= as in create methods

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java [hunk 4]
```diff
@@ -925,10 +935,36 @@ public static ReferencedEnvelope reference(org.opengis.geometry.Envelope env) {
...
+            }
         }
     
+        // emptiness test according to com.vividsolutions.jts.geom.Envelope
+        if(env.getMaximum(0) < env.getMinimum(0))
+            return new ReferencedEnvelope(env.getCoordinateReferenceSystem());
+        
         return new ReferencedEnvelope(env);
     }
+
```
Comment (HendrikPeilke) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: added a method for a Rectangle2D, don't really know if this is needed

File: modules/library/api/src/test/java/org/geotools/geometry/jts/ReferencedEnvelopeTest.java [hunk 5]
```diff
@@ -142,7 +151,7 @@ public void testBoundsEquals2D() {
     @Test
     public void testFactoryMethod() throws Exception {
         try {
-            ReferencedEnvelope bounds = new ReferencedEnvelope( DefaultGeographicCRS.WGS84_3D );
```
Comment (HendrikPeilke) [inline modules/library/api/src/test/java/org/geotools/geometry/jts/ReferencedEnvelopeTest.java]: eliminated an unused variable

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java [hunk 6]
```diff
@@ -21,13 +21,10 @@
 import org.geotools.geometry.DirectPosition2D;
```
Comment (jodygarnett) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: Please update header to (C) 2005-2015

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java (hunk 2 above)
Comment (jodygarnett) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: I thought having a specific BoundingBox method helped one of the compilers not freak out, requiring client code to add casting.

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java (hunk 4 above)
Comment (jodygarnett) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: Can't hurt.

**Comment by jodygarnett** [comment]
Thanks @HendrikPeilke - that is a pretty nice code review you have provided :)

I checked and we do have some contributions from you, can I double check that you have sent OSGeo the paperwork described in [CONTRIBUTING.md](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md)? It can be sent via email these days which is nice.

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java (hunk 6 above)
Comment (HendrikPeilke) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: did it

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java (hunk 2 above)
Comment (HendrikPeilke) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: that was the case for the methods considering a ReferencedEnvelope as a parameter (which is an OGC and a JTS envelope), but the bounding box is only an OGC envelope, so there was no problem.

**Comment by HendrikPeilke** [comment]
Thanks Jody. Yes, I signed and sent the OSGeo paperwork. I got an email back from "Jachym Cepicky <secretary@osgeo.org>" saying "your CLA was filed, thank you". I filled in GeoTools as a "notify project" on the form, but apparantly that did not work as I expected. :)

File: modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java (hunk 4 above)
Comment (HendrikPeilke) [inline modules/library/api/src/main/java/org/geotools/geometry/jts/ReferencedEnvelope.java]: Yes, it is. The constructor is also used in some places, where raster coordinates are transformed to screen coordinates (I don't know if an empty rectangle resulting is really a possible case). I have not problem with removing this method (too much API makes a library more complex and leads towards misunderstandings, that result in workarounds/solutions, which do not use the preferred way of doing things)

**Comment by aaime** [comment]
Hi Jody, had a look, this one seems in order. Do you have further feedback?

**Comment by aaime** [comment]
No more feedback, merging. Just one note for the future, make sure your commit messages are referring to a jira ticket.

**Pull request merged**